File: worldbuilder/plates/kinematics.py
"""
How fast the ground is moving, and what that does where two plates meet.

This is where Euler poles earn their place over drift vectors. A plate turning about an
axis has a different surface velocity everywhere - fast at the equator of its own
rotation, nothing at all at its pole - and that variation is not a detail. It is why one
margin can be pulling apart at one end and grinding sideways at the other, which is what
makes a generated world's geology look like it has reasons.

**Nothing here is stored.** A margin is not classified once and remembered; it is worked
out at the point somebody asks about, from the two plates' motion there.
"""

from dataclasses import dataclass

from ..geometry.sphere import EARTH_RADIUS_M

#: How much of the relative motion must be across the margin rather than along it before
#: the margin is called convergent or divergent rather than transform. Sine of thirty
#: degrees: below that, the plates are mostly sliding past one another.
ACROSS_ENOUGH = 0.5

CONVERGENT = "convergent"
DIVERGENT = "divergent"
TRANSFORM = "transform"


@dataclass(frozen=True)
class Motion:
    """
    What the two plates at a point are doing to each other.

    Attributes:
        margin (Margin): Which plates, and how far away their margin is.
        closing_m_per_myr (float): How fast they approach across the margin. Negative
            means they are separating.
        sliding_m_per_myr (float): How fast they move along it, unsigned.
        kind (str): `convergent`, `divergent` or `transform`.

    """

    margin: object
    closing_m_per_myr: float
    sliding_m_per_myr: float
    kind: str


def surface_velocity(plate, point, radius_m=EARTH_RADIUS_M):
    """
    How fast the ground of one plate is moving at a point.

    Args:
        plate (Plate): The plate.
        point (SpherePoint): Where on the planet.
        radius_m (float, optional): The planet's radius.

    Returns:
        velocity (Vec3): Metres per million years, tangent to the surface.

    Notes:
        The cross product of the rotation vector with the position, scaled by the radius.
        It is automatically tangent to the sphere, and automatically zero at the plate's
        own Euler pole, without either being a special case anybody had to write.

    """
    return plate.angular_velocity().cross(point.vector).scaled(radius_m)


def motion_between(near, far, point, normal, radius_m=EARTH_RADIUS_M):
    """
    What two named plates are doing to each other at a point.

    Args:
        near (Plate): The plate the point is on.
        far (Plate): The plate across the margin.
        point (SpherePoint): Where.
        normal (Vec3): Across the margin, tangent to the surface, pointing towards `near`.
        radius_m (float, optional): The planet's radius.

    Returns:
        motion (Motion): Closing and sliding speeds, and a name for them.

    Notes:
        Split out from `motion_at` so a caller can ask about a margin it has chosen rather
        than the one that happens to be nearest - which is what lets several margins be
        summed instead of one being picked.

    """
    relative = surface_velocity(near, point, radius_m) - surface_velocity(
        far, point, radius_m
    )
    closing = -relative.dot(normal)
    along = relative - normal.scaled(relative.dot(normal))
    sliding = along.length()

    speed = relative.length()
    if speed <= 0.0 or abs(closing) / speed < ACROSS_ENOUGH:
        kind = TRANSFORM
    elif closing > 0.0:
        kind = CONVERGENT
    else:
        kind = DIVERGENT
    return Motion(
        margin=None, closing_m_per_myr=closing, sliding_m_per_myr=sliding, kind=kind
    )


def motion_at(point, plates, radius_m=EARTH_RADIUS_M):
    """
    What is happening across the nearest plate edge, here.

    Args:
        point (SpherePoint): Anywhere on the planet.
        plates (PlateSet): Every plate on the world.
        radius_m (float, optional): The planet's radius.

    Returns:
        motion (Motion or None): None only if there are not two plates to compare.

    Notes:
        The classification is derived and thrown away. Storing it would mean deciding
        once, for a whole margin, what is only true at a point - and a margin that
        converges at one end and slides at the other is the normal case, not an edge one.

    """
    margin = plates.margin_at(point, radius_m)
    if margin.neighbour is None:
        return None

    normal = plates.margin_normal(point, margin)
    if normal is None:
        return None

    relative = surface_velocity(margin.nearest, point, radius_m) - surface_velocity(
        margin.neighbour, point, radius_m
    )

    # The normal points away from the margin, into the nearest plate. So the nearest
    # plate moving *along* it is moving away from the neighbour, and closing is the
    # negative of that.
    closing = -relative.dot(normal)

    along = relative - normal.scaled(relative.dot(normal))
    sliding = along.length()

    speed = relative.length()
    if speed <= 0.0 or abs(closing) / speed < ACROSS_ENOUGH:
        kind = TRANSFORM
    elif closing > 0.0:
        kind = CONVERGENT
    else:
        kind = DIVERGENT

    return Motion(
        margin=margin,
        closing_m_per_myr=closing,
        sliding_m_per_myr=sliding,
        kind=kind,
    )

File: crates/worldbuilder-engine/src/kinematics.rs
//! How fast the ground is moving, and what that does where two plates meet.
//!
//! Ported from `worldbuilder/plates/kinematics.py`. A plate turning about an axis has a
//! different surface velocity everywhere -- fast at the equator of its own rotation,
//! nothing at all at its pole -- and that variation is not a detail. It is why one margin
//! can be pulling apart at one end and grinding sideways at the other, which is what
//! makes a generated world's geology look like it has reasons.
//!
//! **Nothing here is stored.** A margin is not classified once and remembered; it is
//! worked out at the point somebody asks about, from the two plates' motion there.

use crate::plates::{Margin, Plate};
use crate::sphere::SpherePoint;
use crate::vectors::Vec3;

/// How much of the relative motion must be across the margin rather than along it before
/// the margin is called convergent or divergent rather than transform. Sine of thirty
/// degrees: below that, the plates are mostly sliding past one another.
pub const ACROSS_ENOUGH: f64 = 0.5;

/// `convergent`, `divergent` or `transform` -- what a margin's relative motion is doing.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum MarginKind {
    Convergent,
    Divergent,
    Transform,
}

impl MarginKind {
    pub fn as_str(&self) -> &'static str {
        match self {
            MarginKind::Convergent => "convergent",
            MarginKind::Divergent => "divergent",
            MarginKind::Transform => "transform",
        }
    }
}

/// What the two plates at a point are doing to each other.
#[derive(Debug, Clone, Copy)]
pub struct Motion {
    /// Which plates, and how far away their margin is.
    pub margin: Option<Margin>,
    /// How fast they approach across the margin. Negative means they are separating.
    pub closing_m_per_myr: f64,
    /// How fast they move along it, unsigned.
    pub sliding_m_per_myr: f64,
    /// `convergent`, `divergent` or `transform`.
    pub kind: MarginKind,
}

/// How fast the ground of one plate is moving at a point.
///
/// The cross product of the rotation vector with the position, scaled by the radius. It
/// is automatically tangent to the sphere, and automatically zero at the plate's own
/// Euler pole, without either being a special case anybody had to write.
pub fn surface_velocity(plate: &Plate, point: &SpherePoint, radius_m: f64) -> Vec3 {
    plate
        .angular_velocity()
        .cross(&point.vector)
        .scaled(radius_m)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::sphere::EARTH_RADIUS_M;

    /// Slice 1g's `test_plate(index, lat, lon)` fixes the pole at (80.0, 5.0) and the
    /// rate at 0.01, which cannot vary the pole and rate independently. These tests need
    /// both, so this one takes them as explicit parameters.
    fn test_plate_with_pole(
        index: usize,
        seed_lat: f64,
        seed_lon: f64,
        pole_lat: f64,
        pole_lon: f64,
        rate: f64,
    ) -> Plate {
        Plate {
            index,
            seed: SpherePoint::from_latlon(seed_lat, seed_lon),
            euler_pole: SpherePoint::from_latlon(pole_lat, pole_lon),
            rate_rad_per_myr: rate,
        }
    }

    #[test]
    fn a_plate_is_motionless_at_its_own_euler_pole() {
        // Not a special case in the code -- it falls out of the cross product, because
        // the position vector is parallel to the rotation axis there.
        let plate = test_plate_with_pole(0, 0.0, 0.0, 90.0, 0.0, 0.01);
        let at_pole = SpherePoint::from_latlon(90.0, 0.0);
        let v = surface_velocity(&plate, &at_pole, EARTH_RADIUS_M);
        assert!(
            v.length() < 1e-9,
            "velocity at the plate's own Euler pole should vanish, got {}",
            v.length(),
        );
    }

    #[test]
    fn surface_velocity_is_tangent_to_the_sphere() {
        // Also not a special case: a cross product with the position is perpendicular to
        // it.
        let plate = test_plate_with_pole(0, 0.0, 0.0, 90.0, 0.0, 0.01);
        let point = SpherePoint::from_latlon(17.0, 43.0);
        let v = surface_velocity(&plate, &point, EARTH_RADIUS_M);
        assert!(
            v.dot(&point.vector).abs() < 1e-9,
            "velocity must be tangent, dot with position was {}",
            v.dot(&point.vector),
        );
    }

    #[test]
    fn doubling_the_rate_exactly_doubles_the_velocity() {
        // Exact, not approximate, and the reason is worth stating: angular_velocity is
        // linear in the rate; doubling every component multiplies the sum of squares by
        // four; and sqrt(4x) is exactly 2*sqrt(x) in IEEE-754. Scaling by a power of two
        // is exact throughout, so there is no rounding anywhere in this chain.
        let slow = test_plate_with_pole(0, 0.0, 0.0, 90.0, 0.0, 0.01);
        let fast = test_plate_with_pole(0, 0.0, 0.0, 90.0, 0.0, 0.02);
        let point = SpherePoint::from_latlon(17.0, 43.0);
        let a = surface_velocity(&slow, &point, EARTH_RADIUS_M).length();
        let b = surface_velocity(&fast, &point, EARTH_RADIUS_M).length();
        assert_eq!(
            b.to_bits(),
            (2.0 * a).to_bits(),
            "expected exact doubling, got {b} vs {}",
            2.0 * a
        );
    }
}
